### ISSUE TITLE
Feature: Function to determine if density matrix is absolutely PPT

### ISSUE THREAD

**Issue by vprusso** [vprusso/toqito#208]
Similar to the `IsAbsPPT` function from QETLAB ([link](https://github.com/nathanieljohnston/QETLAB/blob/master/IsAbsPPT.m)).

This would also require implementing the `AbsPPTConstraints` or equivalent in toqito as well.

**Comment by sahilsingh2402** [comment]
Please assign this issue to me.

**Comment by purva-thakre** [comment]
@sahilsingh2402 Are you planning to work on this issue?

**Comment by vprusso** [comment]
After discussing this with @sahilsingh2402, they decided to work on other items. I'll unassign them for now.

**Comment by stannum13** [comment]
Please assign this issue to me.

**Comment by vprusso** [comment]
Hi @stannum13, as the project as been getting lots of interest at this time, users sometimes ask to be assigned and then abandon issues (making it confusing whether the issue is still being worked on, etc.). I would suggest you put up a first approach MR initially, and then we can assign it. Does that work for you?

**Comment by stannum13** [comment]
Lovely, that works for me! Let me look into it and get back to you then.

**Comment by RajdeepAher** [comment]
Would love to work on this!

**Comment by luffy-orf** [comment]
@vprusso , Please assign me the issue

**Comment by purva-thakre** [comment]
@luffy-orf I have unassigned you from the issue as the rule for UHACK states that an issue is assigned **only after** a PR is successfully merged.

**Comment by purva-thakre** [comment]
In case someone comes across this issue, it is still open for a bounty. Progress on a PR by a hacker has stalled.

**Comment by Newtech66** [comment]
The QETLAB function allows returning the constraints in a form amenable to usage in `CVX` (`cvxpy` is the Python equivalent). Is this desired as a part of this issue, or is it enough to just implement the checks (which means I don't have to handle returning the conditions as `cvxpy` Constraints)?

**Comment by vprusso** [comment]
> The QETLAB function allows returning the constraints in a form amenable to usage in `CVX` (`cvxpy` is the Python equivalent). Is this desired as a part of this issue, or is it enough to just implement the checks (which means I don't have to handle returning the conditions as `cvxpy` Constraints)?

From what I can see in the QETLAB code, the only thing that is returned is a sentinel value which indicates whether it is/is not absolutely PPT or is indeterminant. The SDP constraints for the dim >= 7 case would need to be enforced, but I'm not sure if this is what you mean (or if I am misunderstanding your comment). Hopefully, that makes sense, but let me know if I'm not responding to the correct question or if I misunderstood what you are asking here!

**Comment by Newtech66** [comment]
I don't know yet, I'm still working through the details of the paper and the implementation. I'll update once I figure it out.

**Comment by vprusso** [comment]
Sounds good, thanks, @Newtech66 !

**Comment by Newtech66** [comment]
> would it be necessary for [this part](https://github.com/nathanieljohnston/QETLAB/blob/f589c35a29e03624ecfd972fd209014799d24862/IsAbsPPT.m#L96) of IsAbsPPT.m

No, what that part is basically doing is, if it's inside a CVX optimization problem, it adds the returned matrices from `AbsPPTConstraints.m` as CVX constraints in an SDP. If it's not inside a CVX optimization problem, that code path is not triggered at all, and it goes to [this line](https://github.com/nathanieljohnston/QETLAB/blob/f589c35a29e03624ecfd972fd209014799d24862/IsAbsPPT.m#L107). Here, if p<=6, it checks all constraints and returns a definite answer. If p>=7, then it checks if any constraint matrix is violated: then the answer is `False`, otherwise it returns an inconclusive result (if you are wondering where it sets that, [it's here](https://github.com/nathanieljohnston/QETLAB/blob/f589c35a29e03624ecfd972fd209014799d24862/IsAbsPPT.m#L35)).

**Comment by vprusso** [comment]
Sounds good, @Newtech66 . Feel free to give me a ping when you'd like a round of reviews if that would be helpful!

**Comment by purva-thakre** [comment]
Closing this so that the UF assigns the bounty to @Newtech66. There are still a few things that need to be resolved in the PR, but it is very close and approved by both maintainers.
